server/service/src/insurance/update/mod.rs

```diff
@@ -0,0 +1,117 @@
+use chrono::{NaiveDate, Utc};
+use repository::{
+    name_insurance_join_row::{
+        InsurancePolicyType, NameInsuranceJoinRow, NameInsuranceJoinRowRepository,
+    },
+    InsuranceProviderRow, InsuranceProviderRowRepository, RepositoryError, StorageConnection,
+};
+
+use crate::{service_provider::ServiceContext, SingleRecordError};
+
+mod generate;
+mod test;
+mod validate;
+
+use generate::{generate, GenerateInput};
+use validate::validate;
+
+#[derive(PartialEq, Debug)]
+pub enum UpdateInsuranceError {
+    InsuranceDoesNotExist,
+    UpdatedRecordNotFound,
+    DatabaseError(RepositoryError),
+}
+
+impl From<RepositoryError> for UpdateInsuranceError {
+    fn from(error: RepositoryError) -> Self {
+        UpdateInsuranceError::DatabaseError(error)
+    }
+}
+
+impl From<SingleRecordError> for UpdateInsuranceError {
+    fn from(error: SingleRecordError) -> Self {
+        use UpdateInsuranceError::*;
+        match error {
+            SingleRecordError::DatabaseError(error) => DatabaseError(error),
+            SingleRecordError::NotFound(_) => UpdatedRecordNotFound,
+        }
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct UpdateInsurance {
+    pub id: String,
+    pub policy_type: Option<InsurancePolicyType>,
+    pub discount_percentage: Option<f64>,
+    pub expiry_date: Option<NaiveDate>,
+    pub is_active: Option<bool>,
+    pub provider_name: Option<String>,
+}
+
+pub fn update_insurance(
+    ctx: &ServiceContext,
+    input: UpdateInsurance,
+) -> Result<NameInsuranceJoinRow, UpdateInsuranceError> {
+    let insurance = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let insurance = validate(connection, &input.clone())?;
+            let new_insurance = generate(GenerateInput {
+                update_input: input.clone(),
+                name_insurance_join_row: insurance,
+            });
+
+            update_insurance_provider(connection, &input, &new_insurance.insurance_provider_id)?;
```

Comment:
I don't think we should be creating any insurance providers here, was a bit confused for a while looking at the code :)

Insurance Providers are configured centrally and synced to open-mSupply. If we did create them in the future they would be from the central server only, not as part of this update method.

Reply:
Misunderstood the requirements, this is good catch! Removing changes for creating insurance providers